### PR TITLE
[Feature][DistDGL] Add NCCL support for range based partitions

### PR DESCRIPTION
**Pull request by nv-dlasalle**
## Description
This adds "range" based ndarray partitions, where a prefix-sum is used to identify which GPU a row in a tensor belongs.

This will enable using the NCCL based SparseOptimizer's and NodeEmbedding in DistDGL, as well as MultiGPUTensor (#3021).

In a subsequent PR, a means of converting a `RangePartitionBook` to this new class will be added.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

## Changes
Add the RangePartition class.

Add unit tests in `tests/cpp/test_partition.cc`.

Add unit tests in `tests/compute/test_nccl.py`.

### CODE REVIEW

File: src/partition/partition_op.h [hunk 1]
```diff
@@ -70,6 +70,74 @@ IdArray MapToGlobalFromRemainder(
     IdArray local_idx,
     int part_id);
 
+/**
+ * @brief Create a permutation that groups indices by the part id.
```
Comment (classicsong) [inline src/partition/partition_op.h:74]: Can you give an example in the docstring?

File: tests/compute/test_nccl.py [hunk 2]
```diff
@@ -52,6 +52,36 @@ def test_nccl_sparse_pull_single():
     exp_rv = F.gather_row(value, req_index)
     assert F.array_equal(rv, exp_rv)
 
+@unittest.skipIf(F._default_context_str == 'cpu', reason="NCCL only runs on GPU.")
```
Comment (classicsong) [inline tests/compute/test_nccl.py:55]: Do we need to skip windows?

**Review by classicsong — APPROVED** [review]
Left some comments.
Overall LGTM.

File: python/dgl/partition.py [hunk 3]
```diff
@@ -409,14 +419,41 @@ class NDArrayPartition(object):
...
     >>> from dgl.partition import NDArrayPartition
     >>> part = NDArrayPartition(g.num_nodes(), num_parts, mode='remainder' )
+
+    A range based partition of a homogenous graph `g`'s nodes, where
+    the nodes are stored in contiguous memory. This converts an existing
+    range based partitioning (e.g. from a
+    dgl.distributed.graph_partition_book.RangePartitionBook)
+    'max_node_map', to an NDArrayPartition 'part'.
+
+    >>> part_range = [0]
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply python/dgl/partition.py:430]: Added `dgl.distributed.graph_partition_book.get_node_partition_from_book` in 
ec8f042 .

File: src/partition/cuda/partition_op.cu [hunk 4]
```diff
@@ -76,6 +76,90 @@ __global__ void _MapGlobalIndexByRemainder(
...
+    const int proc = _SearchRange(
+        range,
+        static_cast<int>(comm_size),
+        static_cast<RangeType>(in[idx]));
+    out[idx] = in[idx] - range[proc];
+  }
+}
+
+template<typename IdType, typename RangeType>
+__global__ void _MapGlobalIndexByRangeKernel(
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply src/partition/cuda/partition_op.cu:233]: Added docs to all kernels and device functions in this file in 
1c97b44

File: tests/compute/test_nccl.py (hunk 2 above)
Comment (nv-dlasalle) [reply tests/compute/test_nccl.py:55]: Do we not enable NCCL on windows builds? Or is there another reason this should skip it?

File: src/partition/partition_op.h (hunk 1 above)
Comment (nv-dlasalle) [reply src/partition/partition_op.h]: Example added in 
3c761c8

File: src/partition/ndarray_partition.cc [hunk 5]
```diff
@@ -105,13 +105,126 @@ class RemainderPartition : public NDArrayPartition {
...
+      const int64_t array_size,
+      const int num_parts,
+      IdArray range) :
+    NDArrayPartition(array_size, num_parts),
+    range_(range),
+    // We also need a copy of the range on the CPU, to compute partition
+    // sizes. We require the input range on the GPU, as if we have multiple
+    // GPUs, we can't know which is the proper one to copy the array to, but we
+    // have only one CPU context, and can safely copy the array to that.
+    range_cpu_(range.CopyTo(DGLContext{kDLCPU, 0})) {
```
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply src/partition/ndarray_partition.cc:120]: It's used when initializing the corresponding local portion of the partitioned tensor, which should be almost all cases (unless the size is gotten from somewhere else like the graph partition book).

**Pull request merged**